### PR TITLE
Recurrence Expander

### PR DESCRIPTION
**Pull request by lightsofapollo**
Add new class to handle the expansion of events based on multiple RRULE's and taking into account RDATE, EXDATE while returning events in order of occurrence.

This is still very early and I am not completely happy with the API I came up with...

My goal is to have something that can expand a limit number of events that can also be persisted into IndexedDB so we can later resume the expansion of events.

This needs more tests and refinement but should give you an idea of what I am trying to do.

### CODE REVIEW

File: lib/ical/helpers.js [hunk 1]
```diff
@@ -29,6 +29,33 @@ ICAL.helpers = {
     };
   },
 
+  binsearchInsert: function(list, seekVal, cmpfunc) {
```
Comment (lightsofapollo) [inline lib/ical/helpers.js:32]: I have taken this from gaia (written by asuth)

File: lib/ical/recur_expansion.js [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    }
+
+    this._ensureDates();
+  }
+
+  function propertyValue(prop) {
+    return prop.data.value[0];
+  }
+
+  RecurExpansion.prototype = {
```
Comment (lightsofapollo) [inline lib/ical/recur_expansion.js]: Not sure about this naming this is what I have been thinking it should be called but my use case is very specific and probably not so much for ICAL as for CALDAV

File: test/mocha/recur_expansion_test.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
+suite('recur_expansion', function() {
```
Comment (lightsofapollo) [inline test/mocha/recur_expansion_test.js:1]: I am sure I missed a few things there are likely bugs and edge cases, I only covered a the case where we have both RDATE and RRULE.

File: lib/ical/recur_expansion.js [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    } else {
+      this.last = this.primaryItem.getFirstPropertyValue('DTSTART');
+    }
+
+    this._ensureDates();
+  }
+
+  function propertyValue(prop) {
+    return prop.data.value[0];
+  }
```
Comment (kewisch) [inline lib/ical/recur_expansion.js:14]: I'd suggest not putting this function between function and prototype of RecurExpansion

File: lib/ical/recur_expansion.js [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+      var i = 0;
+      var len = exceptions.length;
+      var exception;
+      var id;
+
+      for (; i < len; i++) {
+        exception = exceptions[i];
+        id = exception.getFirstPropertyValue('RECURRENCE-ID').toString();
+        //XXX: is exception(s) the right terminology here?
```
Comment (kewisch) [inline lib/ical/recur_expansion.js]: Yes, an exception is an item with RECURRENCE-ID, that holds data about the changes to an instance of a recurring item.

File: lib/ical/recur_expansion.js [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    _ruleDates: null,
+    _exDates: null,
+    _ruleDateInc: 0,
+    _exDateInc: 0,
+
+    last: null,
+
+    /**
+     * Takes an array of components and returns an object
+     * with each component stored by its RECURRENCE-ID (as a string).
```
Comment (kewisch) [inline lib/ical/recur_expansion.js]: If you want full documentation too, missing @param.

File: lib/ical/recur_expansion.js [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+  };
+
+  /**
+   * Iterates through a set of sub components grouping
+   * those that relate to specific recurrences and returning
+   * instances of RecurExpansion.
+   *
+   * XXX: This makes more sense in the CALDAV sense then
+   *      raw ICAL where multiple unrelated events are in
+   *      the same VCALENDAR body
```
Comment (kewisch) [inline lib/ical/recur_expansion.js]: Not quite sure what you mean here? They use the same standard ical, caldav has one event per vcalendar I guess, ical has multiple ones?

File: lib/ical/recur_expansion.js [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+   *
+   * @param {ICAL.icalcomponent} component base component.
+   * @param {String} type subcomponent type (like VEVENT).
+   * @return {Object} object where keys are UID of events and value
+   *                  is an instance of RecurExpansion.
+   */
+  RecurExpansion.fromComponent = function(component, type) {
```
Comment (kewisch) [inline lib/ical/recur_expansion.js]: I'd love to see usage of this function, aside from the testcase.  I agree we should work on the API here and maybe we can also improve performance by somehow linking the recurrence exceptions to the master event. Need to watch out for mem leaks due to cyclic references though.

With "component base component" did you mean VCALENDAR base component, or something else?

**Comment by kewisch** [comment]
I might have missed it, but did you add ical_recurexpansion.js to the Makefile so its packaged?

I guess when you say you want something that expands occurrences, you want something like [this](http://mxr.mozilla.org/comm-central/source/calendar/base/src/calRecurrenceInfo.js#581)?

Expanding them into an indexedDB sounds like a good idea to me, this is definitely an area Lightning could also use performance improvements on.

I think some pretty cool things could be done here. We might want to differ those rules that have an immediate effect if the start date is changed and such that don't. This way we could pre-calculate rules that are commonly used by calculating relative offsets from the start date and then re-applying these calulations on other events. For example if you have a simple yearly rule you want to check if it occurrs this month, you could (aside from more specific optimizations) take the difference between the event's start date and the time range start, then look for a relative offset that comes near this value. These caluclations could stick if they are accessed sufficiently often or there is enough memory.

Anyway, r=philipp for now. I'm about to leave and didn't have the time to think about the API itself, but I at least wanted to get rid of some comments.

File: lib/ical/recur_expansion.js (hunk 4 above)
Comment (lightsofapollo) [inline lib/ical/recur_expansion.js:14]: Ah yes!

File: lib/ical/recur_expansion.js (hunk 8 above)
Comment (lightsofapollo) [inline lib/ical/recur_expansion.js]: component base component <- its amazing what you find reading your own code sometimes =( 
I will clarify the use here by adding a code example (and one to the constructor).

File: lib/ical/recur_expansion.js (hunk 7 above)
Comment (lightsofapollo) [inline lib/ical/recur_expansion.js]: CALDAV makes (what I would say is an optimization) and groups all related events (todo, etc..) in the same VCALENDAR body.
ICAL formats itself don't enforce ordering (I think) so RECURRENCE-ID could in theory appear anywhere in a huge document where as we know each sub-component in the CALDAV response is going to be grouped nicely and either by the primary item or an exception for it.

File: lib/ical/recur_expansion.js (hunk 8 above)
Comment (lightsofapollo) [inline lib/ical/recur_expansion.js]: I will document this further but the idea is something like this:

var calendar; // Top level VCALENDAR from caldav in gaia this is streamed and handled one at a time.
var recurExpand = ICAL.RecurExpansion.fromComponent(calendar)l

if (recurExpand) {
  // three month in the future or 90 for daily cases
  result = recurExpand.next(new Date(2012, 1, 30), 90);
  // push each recurrence over
} else {
  //push single over
}

Comment (kewisch) [inline lib/ical/recur_expansion.js]: I think it might be a bit confusing that the recurrence expansion works on a toplevel vcalendar component, i.e. what should happen for vcalendar components that have multiple unrelated vevents in it?

In Lightning the calIEvent implementation and its friends bundle an item with all its exceptions, which is also why we have calIIcsParser. Maybe we need to either change the parser to do that (this means a bigger API change, and also introduces problems like what to do with parentless items), or at least have some other prototype object that can bundle them.

Comment (lightsofapollo) [inline lib/ical/recur_expansion.js]: This method does the grouping your speaking of and the expansion of recurrences. Events that do not recur will be returned in the final result but not used to create instances of `RecurExpansion`.

The output is an object where the key is the events `UID` and the value is either a component or instance of `RecurExpansion`.
This api works for my purposes with CALDAV but it could be _really_ bad for huge ICAL files with a bunch of unrelated events.

I am mostly happy with what this code does but not with the API, I am back now so I will make another proposal soon. I need to get this over and landed in gaia in the next day or so to meet other commitments so I am willing to land something there first then come back to this to get the API right if we can't find the right solution in the next day or so.

**Comment by lightsofapollo** [comment]
@kewisch I am continuing my work here the majority of the code that has been reviewed is the same but _how_ you use recur expansion is totally changing to the new sax like API we have spoken about.

I have made some changes, that should make using components a lot simpler and is a tie in for the timezone work which I have been wanting to start on (which should now be very minor).

File: lib/ical/component_parser.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
+ICAL.ComponentParser = (function() {
```
Comment (lightsofapollo) [inline lib/ical/component_parser.js:1]: Still a bit rough but I think the idea is sound. I expanded on the idea of "onitem" to "onevent" "ontimezone" etc... so we can add better first class support for top level sub-components.

File: lib/ical/event.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
+ICAL.Event = (function() {
+
+  function Event(component, options) {
```
Comment (lightsofapollo) [inline lib/ical/event.js:3]: Originally, I thought what was Recur Expansion was could be turned into ICAL.Event, after thinking about it more I believe its much better suited to be its own object and consume ICAL.Event(s) once we support more sub-components we can look into abstracting the common pieces like it is done in Lightning. I have tired to be very lightweight here mostly delegating to the component.

I found that I did this is my own code a lot on gaia and its super helpful to have a thin wrapper which we can add functionality to.

File: lib/ical/recur_expansion.js [hunk 11]
```diff
@@ -0,0 +1,361 @@
+ICAL.RecurExpansion = (function() {
```
Comment (lightsofapollo) [inline lib/ical/recur_expansion.js:1]: This is in a "working" state but needs a few more changes:

The getNextOccurances api needs a little more work, I think it should return three things: the event object (the exception when required) start date and end date. End dates are not handled anywhere yet and will require a bit more work. The current support in recur is enough but it assumes that the start date given has the same minute seconds as the original DTSTART. This is not always the cases (in my tests and gaia code) so the RECURRENCE-ID's where not aligned with the output of the recur iterator.

Comment (kewisch) [inline lib/ical/recur_expansion.js:1]: Doesn't it have to be this way anyway (same minute/second), unless its a BYSECOND or BYMINUTE rule?

**Comment by lightsofapollo** [comment]
@kewisch This is ready for a final round, I addressed your review comments and made some more changes to the API. I think I have it right now. I rebased my changes so we have one commit for each component rather then the 11 or so I had previously with my api changes.

Summary:
1. I got rid of the onrecurrenceexception events its much easier to just check event.isRecurrenceException() anyway.
   There may also be cases where people don't care about that and its just extra overhead.
2. simplified ComponentParser per your review comments. My original intent was to detect the callbacks and make
   optimizations when callbacks where not present, that was a bad idea so now there are options to configure the class.
3. RecurExpansion, does not care about exceptions anymore and is slimmer due to the removal of fromComponent.
   getNextOccurrences is replaced with "next" just like recur. ( the old max # and max time can be done outside of ICAL.js)
4. The workflow for expanding recurrences is more modular and (IMO) makes more sense.

Example usage:

``` js


var parser = new ICAL.ComponentParser({ parseEvent: true });
var events = [];
var exceptions = Object.create(null);

parser.onevent = function(item) {
  if (item.isRecurringException()) {
    // there is no magic relation that happens
    // so we need to keep track of exceptions and relate them later on.
    var uid = item.uid;
    if (!(uid in exceptions)) {
      exceptions[uid] = [];
    }
    exceptions[uid].push(item);
  } else {
    events.push(item);
  }
};

parser.oncomplete = function() {
  events.forEach(function(item) {
    var uid = item.uid;
    if (uid in exceptions) {
      exceptions[uid].forEach(item.relateException, item);
    }

    if (item.isRecurring()) {
      var iterator = item.iterator();
      var next;
      // handles, exceptions, multiple RRULES, EXDATE and RDATE
      while ((next = item.next())) {
        // will return exception instead of normal occurrence if available.
        var details = item.getOccurrenceDetails(next);
        // details.startDate
        // details.endDate
        // details.item
        // details.recurrenceId
      }
    } else {
      // normal event
    }
  });
};

parser.process('LONG ICAL STRING!')


```

**Comment by lightsofapollo** [comment]
@kewisch I don't want to make this patch any bigger then it is now.

I am continuing my work on a new branch that will add some recurrence related utilities to event and logic for converting a ICAL.RecurExpansion object into a json object that can be serialized/de-serialized for use with IndexedDB.

**Comment by kewisch** [comment]
Feel free to merge, I think we are done with this round. Thanks for all the great work!
